src/crypto/scrypt.cpp

```diff
@@ -255,10 +256,33 @@ static inline void xor_salsa8(uint32_t B[16], const uint32_t Bx[16])
 
 void scrypt_1024_1_1_256_sp_generic(const char *input, char *output, char *scratchpad)
 {
+	// min 10 max 20. this gives N = 2048 and N = 2097152
+	const unsigned char minNfactor = 10;
+	const unsigned char maxNfactor = 20;
+
+	// epoch times of chain start and current block time
+	int64_t nChainStartTime = 1515002093;
+	int64_t nTimestamp = GetAdjustedTime();
+
+	// n-factor will change every this interval is hit
+	int64_t nChangeInterval = 17280000; // 200 days
+
 	uint8_t B[128];
 	uint32_t X[32];
 	uint32_t *V;
-	uint32_t i, j, k;
+	uint32_t i, j, k, N;
+	unsigned char Nfactor;
+
+	// calculate Nfactor
+	if (nTimeStamp <= nChainStartTime) {
+		Nfactor = minNfactor;
+	} else {
+		int64_t s = nTimestamp - nChainStartTime;
+		int n = s/nChangeInterval + 10;
+		if (n < 0) n = 0;
```

Comment:
Should this be changed to: `if (n < minNfactor) n = minNfactor;`? Also, if we're about to down cast it, shouldn't you technically check for an upper bound?